<!DOCTYPE html>
<html>
<head>
<meta charset="UTF-8">
<title>Insert title here</title>
<link rel="stylesheet"
	href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.0/dist/css/bootstrap.min.css"
	integrity="sha384-B0vP5xmATw1+K9KRQjQERJvTumQW0nPEzvF6L/Z6nronJ3oUOFUFpCjEUQouq2+l"
	crossorigin="anonymous">
<script
	src="https://cdn.jsdelivr.net/npm/bootstrap@4.6.0/dist/js/bootstrap.bundle.min.js"
	integrity="sha384-Piv4xVNRyMGpqkS2by6br4gNJ7DXjqk09RmUpJ8jgGtD7zP9yug3goQfGII0yAns"
	crossorigin="anonymous"></script>
<script src="https://code.jquery.com/jquery-3.5.1.slim.min.js"
	integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj"
	crossorigin="anonymous"></script>
<script
	src="https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js"
	integrity="sha384-9/reFTGAW83EW2RDu2S0VKaIzap3H66lZH81PoYlFhbGU+6BZp6G7niu735Sk7lN"
	crossorigin="anonymous"></script>
<script
	src="https://cdn.jsdelivr.net/npm/bootstrap@4.6.0/dist/js/bootstrap.min.js"
	integrity="sha384-+YQ4JLhjyBLPDQt//I+STsc9iw4uQqACwlvpslubQzn4u2UU2UFM80nGisd026JF"
	crossorigin="anonymous"></script>
<script>  //동기 : 응답을 기다렸다가 실행 / 비동기 : 응답이 올때까지 기다리지 않고 실행
	const fun0 = () => { 
		console.log("작업0 실행");		
		fun1();
 		console.log("작업3 실행");
	};
	
	const fun1 = async() => { //이 함수안에는 비동기 작업이 들어가 있다 / 이함수자체는 비동기함수 = async
		try{ //resolve는 try에서 처리 가능
		var response1 = await fun2(); //fun이 promise를 리턴할때만
		console.log(response1);		
		var response2 = await fun3(); //fun2가 결과값을 줄때까지 기다리겟다
		console.log(response2);
		}catch(response){ //여기 response에는 reject의 객체가 들어옴 / reject 발생시 catch에서 처리 
			if(response.stateCode === 404){
				console.log("페이지가 존재하지 않음");
			}else if(response.stateCode === 500){
				console.log("실행 오류가 발생 했음");
				
			}
		}
	}
	
	const fun2 = () => {
	 	 return new Promise((resolve, reject) => {
	 		 console.log("시간을 요하는 작업1...");
	 		 var result = false;
	 		 if(result){
	 			 resolve({stateCode: 200, data: "작업1 결과"}); //success callback
	 		 }else{
	 			 reject({stateCode: 404}); //failure callback
	 		 }
	 	 });
	};
	
	const fun3 = () => {
	 	 return new Promise((resolve, reject) => {
	 		 console.log("시간을 요하는 작업2...");
	 		 var result = true;
	 		 if(result){
	 			 resolve({stateCode: 200, data: "작업2 결과"}); //success callback
	 		 }else{
	 			 reject({stateCode: 500}); //failure callback
	 		 }
	 	 });
	};
</script>
</head>
<body>
	<div class="card">
		<div class="card-header">동기화 이해하기</div>
		<div class="card-body">
			<button onclick="fun0()" class="btn btn-danger btn-sm">함수 호출</button>
		</div>
	</div>
</body>
</html>